### ISSUE TITLE
Changes to TwistedNames to make it support NAPTR records

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2276]
|<img alt="agatonsax's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| agatonsax reported|
|-|-|
|Trac ID|trac#2276|
|Type|enhancement|
|Created|2006-12-06 16:32:31Z|
|Branch|https://github.com/twisted/twisted/tree/names-naptr-2276-2|

There is presently no support for NAPTR records in either the client or the server code and I needed both. So I have made some small modifications to the code such that TwistedNames would support NAPTR records both on the client and the server side.

Attachments:

* [common.py.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/b4e/b4e131670a970c281e7740526950c0285e9d9fdf/527d4fba0adaacd47f5ea0afb6290f5a66d8ceea.patch) (748 bytes) - added by agatonsax on 2006-12-06 16:33:30Z - Patch to names/common.oy
* [dns.py.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/b4e/b4e131670a970c281e7740526950c0285e9d9fdf/1f31a42495bae0cde6a22d6130d3bd5e0787b719.patch) (4442 bytes) - added by agatonsax on 2006-12-06 16:36:13Z - Patch to names/common.oy
* [test_names.py.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/b4e/b4e131670a970c281e7740526950c0285e9d9fdf/5b836cf61622d98a9a14fe2631014d6bea3181cf.patch) (1179 bytes) - added by agatonsax on 2006-12-06 16:36:55Z - Patch to names/testnames.py
* [patch.1](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/b4e/b4e131670a970c281e7740526950c0285e9d9fdf/03aed9565ddf93f7919ff57d80a38f66838f0a38.1) (4299 bytes) - added by agatonsax on 2008-04-28 10:39:31Z - Added documentation and test
* [patch.2.1](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/b4e/b4e131670a970c281e7740526950c0285e9d9fdf/14e6ebd10012708b850019dcdb63530378ab7ba6.1) (6181 bytes) - added by agatonsax on 2008-05-02 15:13:18Z - Replaced patch.1 with a new version that will fix what's needed

<details><summary>Searchable metadata</summary>

```
trac-id__2276 2276
type__enhancement enhancement
reporter__agatonsax agatonsax
priority__normal normal
milestone__ 
branch__branches_names_naptr_2276_2 branches/names-naptr-2276-2
branch_author__ 
status__closed closed
resolution__fixed fixed
component__names names
keywords__ 
time__1165422751000000 1165422751000000
changetime__1217792997000000 1217792997000000
version__None None
owner__ 
cc__therve
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_2"></a>|@therve commented|
|-|-|

(In [23395]) Branching to 'names-naptr-2276'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_3"></a>|@therve commented|
|-|-|

(In [23396]) Check the patched in, with some small fixes.

Refs [#2276](https://github.com/twisted/twisted/issues/2276)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_4"></a>|@therve set owner to agatonsax|
|-|-|

First, thanks a lot for your contribution. I'm sorry we lost some much time handling it. I think it's already in pretty good shape, but it would need more documentation and tests to land in trunk. Are you willing to help on these topics? Thanks again.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_5"></a>|@therve commented|
|-|-|

(Received by mail:)

> Could do that!
> What would the process be ?

Awesome :). First, it would be great if we could discuss it here. Basically, your patch needs:
 * unit tests for all inserted code.
 * docstrings, for every new methods and attributes.

The classes `Charstr` and `Record_NAPTR` need **direct** tests, not test throught the API. Also, it would be great to test some failure conditions of `lookupNAPTR`.

The following patched should be against the branch I created (names-naptr-2276) for convenience.

Don't hesitate to come back to me if you have more questions.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_7"></a>|@therve set owner to agatonsax|
|-|-|

Great, thanks a lot for the documentation effort. I think we're almost there. Last points:
 * the compDict parameter of Charstr.encode is not tested
 * client.lookupNamingAuthorityPointer is not tested.

Back to you.

**Comment by twisted-trac** [comment]
|<img alt="agatonsax's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_8"></a>|agatonsax commented|
|-|-|

Replying to [therve](#note_7):
> Great, thanks a lot for the documentation effort. I think we're almost there. Last points:
>  * the compDict parameter of Charstr.encode is not tested

compDict should not be used for Charstr, since the compression algorithm specified in RFC1037 only is valid for domain names. So I have 'removed' it.

>  * client.lookupNamingAuthorityPointer is not tested.

It is now.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_9"></a>|@therve commented|
|-|-|

(In [23500]) Apply patch

Refs [#2276](https://github.com/twisted/twisted/issues/2276)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_11"></a>|@therve set owner to agatonsax|
|-|-|

OK I was hoping someone with better DNS knowledge would take a look, but everyone is pretty busy. 

It looks good to me, but as I don't know at all this, I'd like to be convinced it's good. If you can you provide one or two small examples with twisted code and an external thing (dig?), I'll be happy to merge that.

Thanks and sorry for the delay.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun commented|
|-|-|

(In [24371]) Branching to 'names-naptr-2276-2'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun removed owner<br>@exarkun set status to `new`|
|-|-|

Seems good to me as well.  I made some minor changes, including some new tests and a fix for the hashability of `Charstr`.  I'm not sure what kind of example you were hoping for.  How's this?

```
exarkun@charm:/tmp$ cat zone
zone = [SOA('foobar'), NAPTR('foobar', 1, 2, 's', '/foo/bar/i', '')]
exarkun@charm:/tmp$ cat zone
zone = [SOA('foobar'), NAPTR('foobar', 1, 2, 's', '/foo/bar/i', '')]
exarkun@charm:/tmp$ twistd dns --port 2053 --pyzone zone
exarkun@charm:/tmp$ dig -t any foobar @localhost -p 2053

; <<>> DiG 9.3.2-P2.1 <<>> -t any foobar @localhost -p 2053
; (1 server found)
;; global options:  printcmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 26891
;; flags: qr aa rd; QUERY: 1, ANSWER: 2, AUTHORITY: 0, ADDITIONAL: 0

;; QUESTION SECTION:
;foobar.                                IN      ANY

;; ANSWER SECTION:
foobar.                 0       IN      SOA     . . 0 0 0 0 0
foobar.                 0       IN      NAPTR   1 2 "s" "/foo/bar/i" "" .

;; Query time: 1 msec
;; SERVER: 127.0.0.1#2053(127.0.0.1)
;; WHEN: Fri Jul 25 14:24:10 2008
;; MSG SIZE  rcvd: 89

exarkun@charm:/tmp$
```

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_15"></a>|@therve set owner to @exarkun|
|-|-|

Cool, please merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_16"></a>|@exarkun set status to `closed`|
|-|-|

(In [24505]) Merge names-naptr-2276-2

Author: agatonsax, exarkun
Reviewer: exarkun, therve
Fixes: [#2276](https://github.com/twisted/twisted/issues/2276)

Add support for the NAPTR record type to `twisted.names`.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
